### PR TITLE
Add support for details on service logs

### PR DESCRIPTION
**Pull request by dperny**
Adds CLI and client support for details on service logs. CLI component of moby/moby#32996.

### CODE REVIEW

File: client/parse_logs.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/pkg/errors"
+)
+
+// ParseLogDetails takes a details string of key value pairs in the form
+// "k=v,l=w", where the keys and values are url query escaped, and each pair
+// is separated by a comma, returns a map. returns an error if the details
+// string is not in a valid format
+// the exact form of details encoding is implemented in
+// api/server/httputils/write_log_stream.go
+func ParseLogDetails(details string) (map[string]string, error) {
```
Comment (dnephin) [inline client/parse_logs.go]: unit test for this?

Comment (dperny) [inline client/parse_logs.go]: Will do right now

Comment (dperny) [inline client/parse_logs.go]: Actually this is in the wrong place

**Comment by dperny** [comment]
includes #45 because i need tests on docker to pass and they won't without this

**Comment by dperny** [comment]
please ping me directly on slack when you're reviewing this.

File: cli/command/service/logs.go [hunk 2]
```diff
@@ -56,11 +54,13 @@ func newLogsCommand(dockerCli *command.DockerCli) *cobra.Command {
 	// options specific to service logs
 	flags.BoolVar(&opts.noResolve, "no-resolve", false, "Do not map IDs to Names in output")
 	flags.BoolVar(&opts.noTrunc, "no-trunc", false, "Do not truncate output")
+	flags.BoolVar(&opts.raw, "raw", false, "Do not neatly format logs")
```
Comment (thaJeztah) [inline cli/command/service/logs.go]: can you add an API version annotation here?

```go
flags.SetAnnotation("raw", "version", []string{"1.30"})
```

File: cli/command/service/logs.go [hunk 3]
```diff
@@ -56,11 +54,13 @@ func newLogsCommand(dockerCli *command.DockerCli) *cobra.Command {
...
 	flags.BoolVar(&opts.noTaskIDs, "no-task-ids", false, "Do not include task IDs in output")
 	// options identical to container logs
 	flags.BoolVarP(&opts.follow, "follow", "f", false, "Follow log output")
 	flags.StringVar(&opts.since, "since", "", "Show logs since timestamp (e.g. 2013-01-02T13:23:37) or relative (e.g. 42m for 42 minutes)")
 	flags.BoolVarP(&opts.timestamps, "timestamps", "t", false, "Show timestamps")
+	flags.BoolVar(&opts.details, "details", false, "Show extra details provided to logs")
```
Comment (thaJeztah) [inline cli/command/service/logs.go]: can you add an API version annotation here?

```go
flags.SetAnnotation("details", "version", []string{"1.30"})
```

**Comment by dperny** [comment]
rebased onto master. vendored moby at the dependent commit.

**Comment by dperny** [comment]
@thaJeztah fixed.

File: cli/command/service/logs.go [hunk 4]
```diff
@@ -218,40 +231,83 @@ type logWriter struct {
...
-		if i == contextIndex {
-			formatted, err := lw.f.format(lw.ctx, logCtx)
-			if err != nil {
-				return 0, err
-			}
-			output = append(output, []byte(fmt.Sprintf("%s    |", formatted))...)
-		} else {
-			output = append(output, part...)
+		// then sort em
+		sort.Strings(detailKeys)
```
Comment (dnephin) [inline cli/command/service/logs.go]: I don't think you need this. I'm pretty sure you can just sort the `d` slice.

Comment (dperny) [inline cli/command/service/logs.go]: yeah, i think you're right. this is hold-over code from when i was caching the details keys.

Comment (dperny) [inline cli/command/service/logs.go]: fixed

**Comment by dperny** [comment]
@dnephin fixed.

**Comment by aaronlehmann** [comment]
Needs a rebase

**Comment by dperny** [comment]
took the wrong vendor commit on rebase, i force-pushed change but i think gh is broken right now. this should fix it, though.

**Pull request merged**